src/index.ts

```diff
@@ -107,50 +100,37 @@ export class ERC725<Schema extends GenericSchema> {
       ipfsGateway: 'https://cloudflare-ipfs.com/ipfs/',
     };
 
-    // Init options member
     this.options = {
       schema,
       address,
-      providerType: null,
+      provider: this.initializeProvider(provider),
       config: {
         ...defaultConfig,
         ...config,
       },
     };
+  }
 
+  // eslint-disable-next-line class-methods-use-this
+  private initializeProvider(providerOrProviderWrapper) {
     // do not fail on no-provider
-    if (!provider) return;
+    if (!providerOrProviderWrapper) return undefined;
 
-    const givenProvider = provider.provider || provider;
+    switch (true) {
+      case typeof providerOrProviderWrapper.request === 'function':
+        return new EthereumProviderWrapper(providerOrProviderWrapper);
 
-    // CASE: GraphQL provider
+      case !providerOrProviderWrapper.request &&
+        typeof providerOrProviderWrapper.send === 'function':
+        return new Web3ProviderWrapper(providerOrProviderWrapper);
 
-    if (provider.type === 'ApolloClient') {
-      this.options.providerType = ProviderType.GRAPH;
-      this.options.provider = new GraphSource(givenProvider);
+      case providerOrProviderWrapper.type === ProviderTypes.GRAPH_QL:
+        return providerOrProviderWrapper;
 
-      // This checks to see if its a subgraph, since TheGraph subgraphs cannot checksum addresses to store
-      const isSubgraph = givenProvider.link?.options?.uri.includes('/subgraph');
-      if (!isSubgraph && address) {
-        this.options.address = toChecksumAddress(address);
-      }
-
-      // CASE: Ethereum provider
-    } else if (provider.request || provider.type === 'EthereumProvider') {
-      this.options.providerType = ProviderType.ETHEREUM;
-      this.options.provider = new EthereumSource(givenProvider);
-
-      // CASE: Web3 or deprecated ethereum provider
-    } else if (
-      (!provider.request && provider.send) ||
-      provider.type === 'Web3Provider'
-    ) {
-      this.options.providerType = ProviderType.WEB3;
-      this.options.provider = new Web3Source(givenProvider);
-
-      // CASE: Unknown provider
-    } else {
-      throw new Error(`Incorrect or unsupported provider ${givenProvider}`);
+      default:
+        throw new Error(
+          `Incorrect or unsupported provider ${providerOrProviderWrapper}`,
+        );
```

Comment:
I'm not super fan and used to switch on a constant value such as `switch(true)`. Why not use if/if-else/else ?

Reply:
I think it's not very readable (else if) but how about this?
```typescript  
    if (typeof providerOrProviderWrapper.request === 'function')
      return new EthereumProviderWrapper(providerOrProviderWrapper);

    if (
      !providerOrProviderWrapper.request &&
      typeof providerOrProviderWrapper.send === 'function'
    )
      return new Web3ProviderWrapper(providerOrProviderWrapper);

    if (providerOrProviderWrapper.type === ProviderTypes.GRAPH_QL)
      return providerOrProviderWrapper;

    throw new Error(
      `Incorrect or unsupported provider ${providerOrProviderWrapper}`,
    );
```